src/compiler/checker.ts

```diff
@@ -5357,7 +5378,12 @@ namespace ts {
         }
 
         function isContextSensitiveFunctionLikeDeclaration(node: FunctionLikeDeclaration) {
-            return !node.typeParameters && node.parameters.length && !forEach(node.parameters, p => p.type);
+            const areAllParametersUntyped = !forEach(node.parameters, p => p.type);
+            if (node.kind === SyntaxKind.ArrowFunction) {
+                return !node.typeParameters && node.parameters.length && areAllParametersUntyped;
+            }
+            const hasThisType = node.parameters.length && (<Identifier>node.parameters[0].name).text === "this" && node.parameters[0].type;

```

Comment:
Turns out no. It still needs special-casing for arrow functions -- they should be skipped if they don't have arguments; nothing else should since everything else implicitly has `this`. But the function is much simpler now.